compiler/aarch64/codegen/OMRMachine.cpp

```diff
@@ -19,11 +19,457 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
  *******************************************************************************/
 
+#include "aarch64/codegen/OMRMachine.hpp"
+
+#include <stdint.h> // for uint16_t, int32_t, etc
+
 #include "codegen/CodeGenerator.hpp"
 #include "codegen/Machine.hpp"
+#include "codegen/Machine_inlines.hpp"
+#include "codegen/RealRegister.hpp" // for RealRegister, etc
+#include "infra/Assert.hpp" // for TR_ASSERT
 
 OMR::ARM64::Machine::Machine(TR::CodeGenerator *cg) :
-      OMR::Machine(cg)
+      OMR::Machine(NUM_ARM64_GPR, NUM_ARM64_FPR)
+   {
+   _registerFile = (TR::RealRegister **)cg->trMemory()->allocateMemory(sizeof(TR::RealRegister *)*TR::RealRegister::NumRegisters, heapAlloc);
+   self()->initialiseRegisterFile();
+   }
+
+TR::RealRegister *OMR::ARM64::Machine::findBestFreeRegister(TR_RegisterKinds rk,
+                                                            bool excludeGPR0,
+                                                            bool considerUnlatched)
    {
+   int first;
+   int last;
+
+   switch(rk)
+      {
+      case TR_GPR:
+         first = TR::RealRegister::FirstGPR;
+         if (excludeGPR0)
+            first++;
+         last  = TR::RealRegister::LastAssignableGPR;
+         break;
+      case TR_FPR:
+         first = TR::RealRegister::FirstFPR;
+         last  = TR::RealRegister::LastFPR;
+         break;
+   }
+
+   uint32_t bestWeightSoFar = 0xffffffff;
+   TR::RealRegister *freeRegister = NULL;
+   for (int i = first; i <= last; i++)
+      {
+      if ((_registerFile[i]->getState() == TR::RealRegister::Free ||
+           (considerUnlatched &&
+            _registerFile[i]->getState() == TR::RealRegister::Unlatched)) &&
+          _registerFile[i]->getWeight() < bestWeightSoFar)
+         {
+         freeRegister    = _registerFile[i];
+         bestWeightSoFar = freeRegister->getWeight();
+         }
+      }
+   if (freeRegister != NULL && freeRegister->getState() == TR::RealRegister::Unlatched)
+      {
+      freeRegister->setAssignedRegister(NULL);
+      freeRegister->setState(TR::RealRegister::Free);
+      }
+   return freeRegister;
+   }
+
+TR::RealRegister *OMR::ARM64::Machine::freeBestRegister(TR::Instruction *currentInstruction,
+                                                        TR_RegisterKinds rk,
+                                                        TR::RealRegister *forced,
+                                                        bool excludeGPR0)
+   {
+   TR_ASSERT(false, "Not implemented yet.");
+
+   return NULL;
+   }
+
+void OMR::ARM64::Machine::initialiseRegisterFile()
+   {
+   _registerFile[TR::RealRegister::NoReg] = NULL;
+   _registerFile[TR::RealRegister::SpilledReg] = NULL;
+
+   _registerFile[TR::RealRegister::x0] = new (self()->cg()->trHeapMemory()) TR::RealRegister(TR_GPR,
```

Comment:
For all of these initializations you have both `self()->cg()` and `_cg`.  We should be consistent.  I'd say go with `self()->cg()`.